go/flexver/main.go

```diff
@@ -0,0 +1,264 @@
+// versions with differing formats will likely produce nonsensical results (garbage in, garbage out),
+// but best effort is made to correct for basic structural changes, and versions of differing length
+// will be parsed in a logical fashion.
+//
+// See the specification at https://github.com/unascribed/FlexVer/blob/trunk/SPEC.md
+package flexver
+
+import (
+	"errors"
+	"sort"
+	"unicode/utf8"
+)
+
+type componentType uint8
+
+const (
+	componentTypeTextual componentType = iota
+	componentTypeNumeric
+	componentTypePreRelease
+	// componentTypeAppendix (unused; appendix components removed in decompose)
+	componentTypeNull
+)
+
+var nullComponent = component{t: componentTypeNull, value: nil}
+
+type component struct {
+	t     componentType
+	value []rune
+}
+
+// compare determines if this component is less than, equal to or greater than the given component.
+// A negative, zero or positive integer respectively indicates this result.
+func (compA component) compare(compB component) int32 {
+	// If both components are numeric, compare numerically (codepoint-wise)
+	if compA.t == componentTypeNumeric && compB.t == componentTypeNumeric {
+		var i, j int
+		// Ignore leading zeroes
+		for i < len(compA.value) && compA.value[i] == '0' {
+			i++
+		}
+		for j < len(compB.value) && compB.value[j] == '0' {
+			j++
+		}
+		aLen := len(compA.value) - i
+		if aLen != len(compB.value)-j {
+			// Lengths differ; compare by length
+			return int32(aLen - (len(compB.value) - j))
+		}
+		// Compare by digits
+		for ; i < aLen; i, j = i+1, j+1 {
+			if compA.value[i] != compB.value[j] {
+				// Converting to digits is unnecessary (we're subtracting anyway)
+				return compA.value[i] - compB.value[j]
+			}
+		}
+	}
+	// One or both are null
+	if compA.t == componentTypeNull {
+		if compB.t == componentTypePreRelease {
+			return 1
+		}
+		return -1
+	} else if compB.t == componentTypeNull {
+		if compA.t == componentTypePreRelease {
+			return -1
+		}
+		return 1
+	}
+	// Textual comparison (differing type, or both textual/pre-release)
+	minLen := len(compA.value)
+	if len(compB.value) < minLen {
+		minLen = len(compB.value)
+	}
+	for i := 0; i < minLen; i++ {
+		a := compA.value[i]
+		b := compB.value[i]
+		if a != b {
+			// Compare by rune
+			return a - b
+		}
+	}
+	// Compare by length
+	return int32(len(compA.value) - len(compB.value))
+}
+
+// ErrInvalidUTF8 is returned from the *Error functions when a passed string is invalid UTF-8.
+var ErrInvalidUTF8 = errors.New("version string is invalid utf-8")
+
+func isASCIIDigit(r rune) bool {
+	return r >= '0' && r <= '9'
+}
+
+func decompose(str string) ([]component, error) {
+	var out []component
+	i := 0 // (byte index; not same as codepoint index!)
+	for i < len(str) {
+		var cur []rune
+		t := componentTypeTextual
+
+		if str[i] == '+' {
+			break // Ignore appendices
+		} else if str[i] == '-' && len(str) > i+1 {
+			// Mark as pre-release, eat first codepoint
+			t = componentTypePreRelease
+			i++
```

Comment:
The leading - in a prerelease component is *not* ignored

Reply:
Fixed, along with a couple other bugs, in bbec50dfeaa136002c3cc8339bd8336cca14be45